[FLINK-9513] Implement TTL state wrappers factory and serializer for value with TTL

azagrebin opened the pull request:
## What is the purpose of the change

This PR introduces a state factory for wrapping state objects with TTL logic and serialiser of user value with expiration timestamp.
NOTE: This PR is based on #6186 and only last commit 99994dedb9a20244a2addd337617778b17fe8349 makes difference with it and needs review.

## Brief change log

  - abstract state creation in backends with `KeyedStateFactory` interface
  - add `TtlStateFactory`
  - add `CompositeSerializer`

## Verifying this change

This change is a trivial addition without any test coverage in this PR and should be covered together with TTL feature activation by final integration and e2e tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (yes)
  - The runtime per-record code paths (performance sensitive): (not yet)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (not yet)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (not applicable at the moment)

sihuazhou commented on flink-core/src/main/java/org/apache/flink/api/common/typeutils/CompositeSerializer.java:
```diff
@@ -0,0 +1,204 @@
+
+	@Override
+	public T deserialize(DataInputView source) throws IOException {
+		List originalValues = new ArrayList();
```
I would suggest to give a init size for `originalValues`, e.g. `List originalValues = new ArrayList(originalSerializers.size());`.

sihuazhou commented on flink-core/src/main/java/org/apache/flink/api/common/typeutils/CompositeSerializer.java:
```diff
@@ -0,0 +1,204 @@
+
+	@Override
+	public T deserialize(T reuse, DataInputView source) throws IOException {
+		List originalValues = new ArrayList();
```
Again, a init size for the List could be given.

sihuazhou commented on flink-core/src/main/java/org/apache/flink/api/common/typeutils/CompositeSerializer.java:
```diff
@@ -0,0 +1,204 @@
+
+			if (previousSerializersAndConfigs.size() == originalSerializers.size()) {
+
+				List<TypeSerializer> convertSerializers = new ArrayList<>();
```
Again, could give a init size for the list.

sihuazhou commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/AbstractKeyedStateBackend.java:
```diff
@@ -48,7 +48,8 @@
 	Closeable,
-	CheckpointListener {
+	CheckpointListener,
+	KeyedStateFactory{
```
I think this seems to miss a space ` `.

sihuazhou commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/ttl/TtlStateFactory.java:
```diff
@@ -0,0 +1,207 @@
+	private final TtlConfig ttlConfig;
+	private final TtlTimeProvider timeProvider;
+
+	private TtlStateFactory(KeyedStateFactory originalStateFactory, TtlConfig ttlConfig, TtlTimeProvider timeProvider) {
```
Would be better to check the args are not null, or simply use the `@Nonnull` annotation.

azagrebin replied:
I added checks in `createStateAndWrapWithTtlIfEnabled`, the only user of this private constructor.

sihuazhou commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/ttl/TtlStateFactory.java:
```diff
@@ -0,0 +1,207 @@
+				.createState(namespaceSerializer, stateDesc);
+	}
+
+	private final Map<Class<? extends StateDescriptor>, StateFactory> stateFactories;
```
Why this couldn't be static?

azagrebin replied:
It could be static. The idea was to create `TtlStateFactory` object for each call of `createStateAndWrapWithTtlIfEnabled`. I think creating of state object will not happen often, only at init. The object contains non static references to originalStateFactory, ttlConfig, timeProvider to make `createXxxState` methods less verbose. `stateFactories` contains reference to this object in `createStateFactories`, so it cannot be static unless everything is static and `createXxxState` are more verbose and accept additional parameters: originalStateFactory, ttlConfig, timeProvider.

sihuazhou replied:
I see, but this looks a bit weird, it looks like that the `stateFactories` is only used for looking up the corresponding state factory for the `stateDesc`, and we need to firstly create it every time when calling the `createStateAndWrapWithTtlIfEnabled()`, the flow of the `createStateAndWrapWithTtlIfEnabled()` looks like:

- create a map of `state factory` (stateFactories).
- use the stateFactories to look up the corresponding state factory for the `stateDesc`.
- ...

In that case, maybe use the `switch case` to find the corresponding state factory is better, at lest we don't need to firstly create a map of `state factory` for every call  this way. What do you think?

azagrebin replied:
I think creating the map is not a big deal because it will happen on init without big pressure on GC. Also map is more declarative and flexible if new state needs to be added then just the map needs to be modified but not the method. Or if needed in future, it can be easier refactored to build the mapping somewhere else and inject it into factory.

sihuazhou commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/ttl/TtlStateFactory.java:
```diff
@@ -0,0 +1,207 @@
+
+		@Override
+		@SuppressWarnings("unchecked")
+		protected TtlValue<T> composeValue(List values) {
```
Should we check that `values != null && values.size() == 2 `?

sihuazhou commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/ttl/TtlStateFactory.java:
```diff
@@ -0,0 +1,207 @@
+
+		@Override
+		@SuppressWarnings("unchecked")
+		protected CompositeSerializer<TtlValue<T>> createSerializerInstance(List<TypeSerializer> typeSerializers) {
```
Should we check that `typeSerializers != null && typeSerializers.size() == 1`?

sihuazhou commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/ttl/TtlStateFactory.java:
```diff
@@ -0,0 +1,207 @@
+		}
+
+		@Override
+		protected List decomposeValue(TtlValue<T> v) {
```
Should we check `v != null`?

sihuazhou commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/ttl/TtlStateFactory.java:
```diff
@@ -0,0 +1,207 @@
+	}
+
+	private static class TtlSerializer<T> extends CompositeSerializer<TtlValue<T>> {
+		TtlSerializer(TypeSerializer<T> userValueSerializer) {
```
The `userValueSerializer` seems can't be null?

sihuazhou commented on flink-core/src/main/java/org/apache/flink/api/common/typeutils/CompositeSerializer.java:
```diff
@@ -0,0 +1,204 @@
+	public void serialize(T record, DataOutputView target) throws IOException {
+		List originalValues = decomposeValueInternal(record);
+		for (int i = 0; i < originalSerializers.size(); i++) {
+			originalSerializers.get(i).serialize(originalValues.get(i), target);
```
If `originalSerializers` or the `originalValues` is a type of something like `LinkedList`, then the `originalSerializers.get(i)` and `originalValues.get(i)` will get a very poor performance. I think we might should use the `iterator` here.

sihuazhou commented on flink-core/src/main/java/org/apache/flink/api/common/typeutils/CompositeSerializer.java:
```diff
@@ -0,0 +1,204 @@
+
+	@Override
+	public int getLength() {
+		return originalSerializers.stream().allMatch(s -> s.getLength() >= 0) ?
```
nit: if we don't use the high level API, we could do this in a single loop with a better performance.

StefanRRichter replied:
+1

sihuazhou commented on flink-core/src/main/java/org/apache/flink/api/common/typeutils/CompositeSerializer.java:
```diff
@@ -0,0 +1,204 @@
+		List originalReuseValues = decomposeValueInternal(reuse);
+		return composeValueInternal(
+			IntStream.range(0, originalSerializers.size())
+				.mapToObj(i -> originalSerializers.get(i).copy(originalFromValues.get(i), originalReuseValues.get(i)))
```
I would suggest to use the `Iterator` instead of `xxx.get(i)` here, because if the `originalSerializers` (or the other instances) is type of LinkedList the performance could be bad.

StefanRRichter replied:
+1

sihuazhou commented on flink-core/src/main/java/org/apache/flink/api/common/typeutils/CompositeSerializer.java:
```diff
@@ -0,0 +1,204 @@
+		List originalValues = decomposeValueInternal(from);
+		return composeValueInternal(
+			IntStream.range(0, originalSerializers.size())
+				.mapToObj(i -> originalSerializers.get(i).copy(originalValues.get(i)))
```
Same here might result in a bad performance because of the random access of the list.

StefanRRichter replied:
I would also suggest to consider lower-level constructs for potentially hot methods like this one because this api can easily introduce (non obvious) performance regression.

sihuazhou commented on flink-core/src/main/java/org/apache/flink/api/common/typeutils/CompositeSerializer.java:
```diff
@@ -0,0 +1,204 @@
+		List originalValues = new ArrayList();
+		List originalReuseValues = decomposeValueInternal(reuse);
+		for (int i = 0; i < originalSerializers.size(); i++) {
+			originalValues.add(originalSerializers.get(i).deserialize(originalReuseValues.get(i), source));
```
Same here.

sihuazhou commented on flink-core/src/main/java/org/apache/flink/api/common/typeutils/CompositeSerializer.java:
```diff
@@ -0,0 +1,204 @@
+						f.f0,
+						UnloadableDummyTypeSerializer.class,
+						f.f1,
+						originalSerializers.get(i));
```
Same here.

StefanRRichter commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/KeyedStateFactory.java:
```diff
@@ -0,0 +1,41 @@
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+
+/** This factory produces concrete state objects in backends. */
+public interface KeyedStateFactory {
```
I think this interface can go into the file of `KeyedStateBackend` because it belongs into that context.

StefanRRichter commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/ttl/TtlStateFactory.java:
```diff
@@ -0,0 +1,207 @@
+		).collect(Collectors.toMap(t -> t.f0, t -> t.f1));
+	}
+
+	private interface StateFactory {
```
Do we really need this interface? It looks identical to `KeyedStateFactory` and seems that interface would fit here as well?

azagrebin replied:
Leftover from previous iteration 👍

StefanRRichter commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/ttl/TtlStateFactory.java:
```diff
@@ -0,0 +1,207 @@
+		StateDescriptor<S, SV> stateDesc) throws Exception {
+		SV defVal = stateDesc.getDefaultValue();
+		TtlValue<SV> ttlDefVal = defVal == null ? null : new TtlValue<>(defVal, Long.MAX_VALUE);
+		ValueStateDescriptor<TtlValue<SV>> ttlDescriptor = new ValueStateDescriptor<>(
```
Why are we using the deprecated constructor instead of following the suggested replacement of handling `null` values properly ourselves?

azagrebin replied:
agree, I think we may actually not expose default value for value state with TTL or new descriptor builders in general

StefanRRichter commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/ttl/TtlStateFactory.java:
```diff
@@ -0,0 +1,207 @@
+	private <T, N, SV, S extends State, IS extends S> IS createFoldingState(
+		TypeSerializer<N> namespaceSerializer,
+		StateDescriptor<S, SV> stateDesc) throws Exception {
+		FoldingStateDescriptor<T, SV> foldingStateDescriptor = (FoldingStateDescriptor<T, SV>) stateDesc;
```
Could suppress deprecation warning on the because we need to support it.

StefanRRichter commented on flink-core/src/main/java/org/apache/flink/api/common/typeutils/CompositeSerializer.java:
```diff
@@ -0,0 +1,204 @@
+ */
+@SuppressWarnings("unchecked")
+public abstract class CompositeSerializer<T> extends TypeSerializer<T> {
+	private final List<TypeSerializer> originalSerializers;
```
Why is this class using a lot of raw types instead of wildcards? E.g. why `List<TypeSerializer>`instead of `List<TypeSerializer<?>>`

StefanRRichter commented on flink-core/src/main/java/org/apache/flink/api/common/typeutils/CompositeSerializer.java:
```diff
@@ -0,0 +1,204 @@
+		this.originalSerializers = originalSerializers;
+	}
+
+	protected abstract T composeValue(List values);
```
Also here, list is just used as raw type. I think also the abstract methods could use some documentation for people that would like to create new subclasses in the future.

StefanRRichter commented on flink-core/src/main/java/org/apache/flink/api/common/typeutils/CompositeSerializer.java:
```diff
@@ -0,0 +1,204 @@
+ * @param <T> type of custom serialized value
+ */
+@SuppressWarnings("unchecked")
+public abstract class CompositeSerializer<T> extends TypeSerializer<T> {
```
You could simply have tests for all the new serializer by extending `SerializerTestBase`, which I would recommend to do to catch mistakes.

StefanRRichter commented:
Thanks for the nice contribution. I had some comments inline. In particular three points about the serializer. I would suggest to avoid the use of raw types. I would also suggest to avoid using streaming API at least methods that can appear in hot loops (mainly copy, de/serialize) for performance reasons. I think the imperative style code will not even be (much) longer in those cases. Last, I suggest to always test new serializers via the `SerializerTestBase` because this catches many problems with little effort.

azagrebin commented:
@StefanRRichter @sihuazhou 
Thanks guys for the helpful review. 
I refactored the `CompositeSerializer` to rely rather on loops than streams and added tests for it.
Please, have a look.

cc @twalthr 
Maybe, the implementation of `CompositeSerializer` in this PR could be potentially interesting for other complex types, e.g. tuples or pojos allowing code reuse where it is possible without performance loss.

azagrebin commented:
Regarding CI, seems to be unrelated test `YARNSessionCapacitySchedulerITCase`, [passed in my CI](https://travis-ci.org/azagrebin/flink/builds/398352328)

StefanRRichter commented on flink-core/src/main/java/org/apache/flink/api/common/typeutils/CompositeSerializer.java:
```diff
@@ -0,0 +1,258 @@
+		boolean stateful = false;
+		for (int index = 0; index < fieldSerializers.length; index++) {
+			duplicatedSerializers[index] = fieldSerializers[index].duplicate();
+			if (fieldSerializers[index] != duplicatedSerializers[index]) {
```
I wonder if we need to do these checks every time `duplicate()` is called? We could check it once, remember if all field serializer are stateless and from that point return `this` immediately.

StefanRRichter commented on flink-core/src/main/java/org/apache/flink/api/common/typeutils/CompositeSerializer.java:
```diff
@@ -0,0 +1,258 @@
+
+	@Override
+	public boolean isImmutableType() {
+		for (TypeSerializer<Object> fieldSerializer : fieldSerializers) {
```
Why not compute this once in the constructor and remember in a boolean flag??

StefanRRichter replied:
We can compute many things like length, immutability, etc already in the constructor. Statelessness is the one thing that we might want to figure out and remember on the first attempt.

StefanRRichter commented on flink-core/src/main/java/org/apache/flink/api/common/typeutils/CompositeSerializer.java:
```diff
@@ -0,0 +1,258 @@
+		for (int index = 0; index < fieldSerializers.length; index++) {
+			fields[index] = fieldSerializers[index].copy(getField(from, index));
+		}
+		return createInstance(fields);
```
This method can return `from` if `isImmutableType()`

StefanRRichter commented on flink-core/src/main/java/org/apache/flink/api/common/typeutils/CompositeSerializer.java:
```diff
@@ -0,0 +1,258 @@
+		for (int index = 0; index < fieldSerializers.length; index++) {
+			fields[index] = fieldSerializers[index].copy(getField(from, index), getField(reuse, index));
+		}
+		return fromFields(fields, reuse);
```
Same here, could return `this` for immutable types.

StefanRRichter commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/ttl/TtlStateFactory.java:
```diff
@@ -0,0 +1,210 @@
+/**
+ * This state factory wraps state objects, produced by backends, with TTL logic.
+ */
+@SuppressWarnings("unchecked")
```
I would not suppress warnings in the scope of a full class, better more fine grained on methods.

StefanRRichter commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/ttl/TtlStateFactory.java:
```diff
@@ -0,0 +1,210 @@
+ * This state factory wraps state objects, produced by backends, with TTL logic.
+ */
+@SuppressWarnings("unchecked")
+public class TtlStateFactory {
```
It might make sense to also have a test when we introduce a new class that provides some kind of "service", e.g. to check that all the types are correctly mapped and also prevent that somebody breaks the mapping by accident.

StefanRRichter commented:
I had a few more comments, in particular some improvements for the new serializer. I think when those are addressed this is good to merge.

StefanRRichter commented on flink-core/src/main/java/org/apache/flink/api/common/typeutils/CompositeSerializer.java:
```diff
@@ -0,0 +1,258 @@
+	private static final long serialVersionUID = 1L;
+
+	protected final TypeSerializer<Object>[] fieldSerializers;
+	final boolean isImmutableTargetType;
```
I think in Java code style, a boolean field name should not be prefixed with `is...`, only the getter should be prefixed with `is...`

StefanRRichter commented on flink-core/src/main/java/org/apache/flink/api/common/typeutils/CompositeSerializer.java:
```diff
@@ -0,0 +1,275 @@
+		this.fieldSerializers = (TypeSerializer<Object>[]) fieldSerializers;
+		this.length = calcLength();
+		this.stateful = isStateful();
+		this.hashCode = Arrays.hashCode(fieldSerializers);
```
I think up to this point, the code is iterating `fieldSerializers` 5 times (null checks, immutable check, length calc, stateful check, and hash code computation. It could be done in one iteration, but since this method should typically not be called in hot loops, this is an optional improvement.

StefanRRichter commented on flink-core/src/main/java/org/apache/flink/api/common/typeutils/CompositeSerializer.java:
```diff
@@ -0,0 +1,275 @@
+	}
+
+	private boolean isStateful() {
+		TypeSerializer[] duplicatedSerializers = duplicateFieldSerializers();
```
The flag for `isStateful()` is the only one that I suggested as a candidate for lazy init when `duplicate()` is called for the first time. Reason is that duplicating some types of inner serializers can sometimes be a bit expensive. But again, I feel that this can also be changed in followup work, if needed.

StefanRRichter commented on flink-core/src/main/java/org/apache/flink/api/common/typeutils/CompositeSerializer.java:
```diff
@@ -0,0 +1,275 @@
+
+	@Override
+	public CompositeSerializer<T> duplicate() {
+		return stateful ? createSerializerInstance(duplicateFieldSerializers()) : this;
```
Another small point here for `createSerializerInstance(...)`: we have no (non-public) constructor that can also take all boolean flags, length, and (maybe) hash directly. So if we copy the serializer, I guess it always goes through the whole process again to figure this out, but we could just copy it from the previous instance.

StefanRRichter commented:
Had a few more comments, but they all are basically optimizations. I leave it up to you if you still want to address all or some of them. Please let me know. Otherwise, we can merge this. 👍

StefanRRichter commented on flink-core/src/main/java/org/apache/flink/api/common/typeutils/CompositeSerializer.java:
```diff
@@ -272,4 +254,60 @@ public int getVersion() {
 	}
+
+	/** This class holds composite serializer parameters which can be precomputed in advanced for better performance. */
+	protected static class PrecomputedParameters implements Serializable {
```
If this is serializable, we should add a version uid.

StefanRRichter commented on flink-core/src/main/java/org/apache/flink/api/common/typeutils/CompositeSerializer.java:
```diff
@@ -272,4 +254,60 @@ public int getVersion() {
+		/** Whether any field serializer is stateful. */
+		final boolean stateful;
+
+		final int hashCode;
```
I wonder if this should be `transient` in a serializable class, the hash code could be based on object identity.

StefanRRichter commented:
LGTM 👍 merging.

StefanRRichter commented:
I found there is still a small issue with the equals/hashCode but will just fix it before merging.
